scarlet/operators_pybind11.cc

```diff
@@ -21,66 +24,69 @@ void prox_monotonic(
   }
 }
 
+template <typename T, typename M, typename V>
 void prox_weighted_monotonic(
     // Fast implementation of weighted monotonicity constraint
-    py::array_t<double> &X,
+    Eigen::Ref<V> flat_img,
     double const &step,
-    py::array_t<double> &weights,
-    std::vector<int> const &offsets,
-    std::vector<int> const &dist_idx,
-    double const &thresh
+    Eigen::Ref<const M> weights,
+    Eigen::Ref<const EigenVector> offsets,
+    Eigen::Ref<const EigenVector> dist_idx,
+    T const &thresh
 ){
-    auto x = X.mutable_unchecked<1>();
-    auto w = weights.mutable_unchecked<2>();
-
     // Start at the center of the image and set each pixel to the minimum
     // between itself and its reference pixel (which is closer to the peak)
-    for(auto &didx: dist_idx){
-        double ref_flux = 0;
-        for(std::size_t i=0; i<offsets.size(); i++){
-            if(w(i,didx)>0){
+    for(int d=0; d<dist_idx.size(); d++){
+        int didx = dist_idx(d);
+        T ref_flux = 0;
+        for(int i=0; i<offsets.size(); i++){
+            if(weights(i,didx)>0){
                 int nidx = offsets[i] + didx;
-                ref_flux += x(nidx) * w(i, didx);
+                ref_flux += flat_img(nidx) * weights(i, didx);
             }
         }
-        x(didx) = std::min(x(didx), ref_flux*(1-thresh));
+        flat_img(didx) = std::min(flat_img(didx), ref_flux*(1-thresh));
     }
 }
 
-// This function is currently unused but might be useful to implement
-// and speed up PSF covolution
-void apply_filter(
-    py::array_t<double> &X,
-    py::array_t<double> &weights,
-    std::vector<std::vector<std::vector<ssize_t>>> &slices,
-    std::vector<std::vector<std::vector<ssize_t>>> &inv_slices,
-    py::array_t<double> &result
+// Apply a filter to an image
+template <typename M, typename V>
+M apply_filter(
+    Eigen::Ref<const M> image,
+    Eigen::Ref<const V> values,
+    Eigen::Ref<const EigenVector> y_start,
+    Eigen::Ref<const EigenVector> y_end,
+    Eigen::Ref<const EigenVector> x_start,
+    Eigen::Ref<const EigenVector> x_end
```

Comment:
It probably doesn't matter, but you might be able to make this go a bit faster by having this take an existing NumPy array (via `Eigen::Ref`) as an output parameter - returning a true `Eigen::Matrix` object will involve a deep copy when converting to NumPy.

Reply:
Good call, it is slightly faster when using PSF deconvolution now.